state/application.go

```diff
@@ -1676,6 +1676,14 @@ func (a *Application) SetCharm(cfg SetCharmConfig) (err error) {
 		}
 
 		if cfg.CharmOrigin != nil {
+			origin := cfg.CharmOrigin
```

Comment:
Arguably it would be clearly to do `origin := *cfg.CharmOrigin` to dereference the pointer here, as we know we're dealing with a non-nil pointer inside this `if`.